### ISSUE TITLE
Using a getter to "populate" (return a reference from) a sub-doc causes validator to run from the getter and not the sub-doc location.

### ISSUE THREAD

**Issue by JasonCust** [Automattic/mongoose#4199]
Granted this might be a "too clever" use case but it provides interesting functionality if done correctly.

Given the schemas:
```javascript
var personSchema = new Schema({
    name: {
        type: String,
        validate: [{
            validator: function logIt(name) {
                console.log('NAME VALIDATE', name, this);
                return true;
            }
        }]
    }
});

var teamSchema = new Schema({
    name: { type: String },
    members: {
        type: [personSchema],
        validate: [{
            validator: function logIt(members) {
                console.log('MEMBERS VALIDATE');
                return true;
            }
        }]
    },
    captain: {
        type: Schema.Types.ObjectId,
        // NOTE: this is the special "getter-reference". It "populates" the value with a
        //  reference for the selected sub-doc array document.
        get: function getReferencedMember(personId) {
            console.log('CAPTAIN GET', personId);
            return this.members.id(personId);
        }
    }
}, {
  toObject: { getters: true, virtuals: true },
  toJSON: { getters: true, virtuals: true }
});
```
And running this test code:

```javascript
var Team = connection.model('Team', teamSchema);
var team = new Team({ name: 'All-stars' });
var members = [
    {name: 'John'},
    {name: 'Joe'},
    {name: 'Jake'},
    {name: 'Jane'},
    {name: 'Jess'},
    {name: 'Jill'}
];

members.forEach(function(member) {
  team.members.push(member);
});
team.captain = team.members[0]._id;

console.log('SAVING TEAM');
team.save().then(function(savedTeam) {
    console.log('CREATED TEAM');

    return Team.findById(savedTeam.id).exec();
}).then(function(foundTeam) {
    console.log('FOUND TEAM');

    team.members[0].name = 'Bill';
    // NOTE: being able to set the name through the reference (as shown below)
    //    is part of the of the intended reason for using the "getter reference" but
    //    will result in the same output obviously.
    // team.captain.name = 'Bill';

    console.log('MODIFIED PATHS', team.modifiedPaths(), team.captain.modifiedPaths());

    console.log('SAVING TEAM');
    return team.save();
}).catch(function(err) {
    console.err(err);
});
```
Will output:

```shell
SAVING TEAM
MEMBERS VALIDATE
NAME VALIDATE John { name: 'John', _id: 575080cd6a3a45056fd64569 }
NAME VALIDATE Joe { name: 'Joe', _id: 575080cd6a3a45056fd6456a }
NAME VALIDATE Jake { name: 'Jake', _id: 575080cd6a3a45056fd6456b }
NAME VALIDATE Jane { name: 'Jane', _id: 575080cd6a3a45056fd6456c }
NAME VALIDATE Jess { name: 'Jess', _id: 575080cd6a3a45056fd6456d }
NAME VALIDATE Jill { name: 'Jill', _id: 575080cd6a3a45056fd6456e }
CAPTAIN GET 575080cd6a3a45056fd64569
CREATED TEAM
FOUND TEAM
CAPTAIN GET 575080cd6a3a45056fd64569
MODIFIED PATHS [ 'members', 'members.0', 'members.0.name' ] [ 'name' ]
SAVING TEAM
NAME VALIDATE Bill { members: 
   [ { _id: 575080cd6a3a45056fd64569, name: 'Bill' },
     { _id: 575080cd6a3a45056fd6456a, name: 'Joe' },
     { _id: 575080cd6a3a45056fd6456b, name: 'Jake' },
     { _id: 575080cd6a3a45056fd6456c, name: 'Jane' },
     { _id: 575080cd6a3a45056fd6456d, name: 'Jess' },
     { _id: 575080cd6a3a45056fd6456e, name: 'Jill' } ],
  _id: 575080cd6a3a45056fd64568,
  name: 'All-stars',
  captain: 575080cd6a3a45056fd64569,
  __v: 0 }
CAPTAIN GET 575080cd6a3a45056fd64569
```

A couple of questions:
- Shouldn't the name validator run from the sub-doc as it has a modified path?
- Why is the name validator run from the location of the "getter-reference"?

Tested on Mongoose 4.4.7 with Mongo 3.2.5.

**Comment by JasonCust** [comment]
Forgot to include that it was tested on Mongoose 4.4.7 with Mongo 3.2.5.

**Comment by JasonCust** [comment]
Confirmed the same output on Mongoose 4.4.19.

**Comment by vkarpov15** [comment]
1) Yeah that's a limitation with how mongoose works currently, see #3771 
2) I'm not sure what you mean by this question, can you please clarify?

**Comment by JasonCust** [comment]
1) They might be related.
2) I was confused as to why initially during create (using `push`) the `this` for the `personSchema.name` validator was the sub-doc being validated and then afterwords `this` was the parent document. Initially I thought it was due to the getter for `teamSchema.captain` but after a little more debugging I see all updates to `teamSchema.members` sub-docs are using the expected `this` (the parent document).

**Comment by vkarpov15** [comment]
Hmm yeah point 2 is a little strange, but will be fixed when #3771 is fixed so I'm gonna mark this as closed. Feel free to re-open if you think this is unresolved.
